drivers/syslog/syslog_channel.c

```diff
@@ -44,9 +44,7 @@
  * Private Function Prototypes
  ****************************************************************************/
 
-#if defined(CONFIG_ARCH_LOWPUTC)
-#  define HAVE_LOWPUTC
-#elif !defined(CONFIG_RAMLOG_SYSLOG) && !defined(CONFIG_SYSLOG_RPMSG)
+#if !defined(CONFIG_RAMLOG_SYSLOG) && !defined(CONFIG_SYSLOG_RPMSG)
 #  define NEED_LOWPUTC
 #endif
```

Comment:
I'm confused about this change in this PR.

Reply:
The function `sc_putc()` has changed prototype, so a pointer to the device is passed:
```c
typedef CODE int (*syslog_putc_t)(FAR void *dev, int ch);
```

Thus `up_putc()` is incompatible. So I use `syslog_default_putc()` to translate the arguments from the new `sc_putc()` to the existing `up_putc()`.

Notice the new argument in `drivers/syslog/syslog_channel.c`, line 112.

Reply:
Got it, much clearer now

Reply:
Just to note, my intention is to add a new channel in the future, that it will encapsulate ARM ITM, up_putc, rpmsg etc, so even these interfaces follow the common structure, and use a proper `struct syslog_dev_s` instead of this ungly `void *` in `struct syslog_channel_s`.

I believe that it will simplify the code more.

Reply:
I agree, that is a good goal

Reply:
@fjpanag it looks that syslog channel become a normal driver. Should we follow other drivers convention? e.g. spi_dev_s/spi_ops_s or rtc_lowerhalf_s/rtc_ops_s, so other people will be more familiar with your change.

Reply:
Yes, this is what I had in mind, but you expressed it in a better way. Having upper/lower drivers for syslog too.

However this means that new lower drivers are to be created (as mentioned above), files may need to be moved etc.

I would prefer to finalize this migration in a later PR, after the lower-hald driver interface is finalized (`sc_close()` is still to be introduced, thread-safety protections are to be added and more).

What do you think? Shall we keep this part for the moment?

Reply:
Something that I am also thinking is to completely drop `sc_putc()` and require on every lower-half driver to implement an `sc_write()` only, with the requirement that the driver will perform the write atomically. If the driver cannot actually perform a write, it can iterate internally in a protection section of the code.

Reply:
> Yes, this is what I had in mind, but you expressed it in a better way. Having upper/lower drivers for syslog too.
> 
> However this means that new lower drivers are to be created (as mentioned above), files may need to be moved etc.
> 

Not to much to become upper/lower:

1. Rename syslog_channel_s to syslog_channel_ops_s
2. Add new struct syslog_channel_s which contain the pointer of syslog_channel_ops_s
3. The callback in syslog_channel_ops_s accept the pointer of syslog_channel_s as the new first argument

> I would prefer to finalize this migration in a later PR, after the lower-half driver interface is finalized (`sc_close()` is still to be introduced, thread-safety protections are to be added and more).
> 

The more callback like sc_close could be added in later if required

> What do you think? Shall we keep this part for the moment?

I am fine which one come first, my major concern is to reduce your implmentation cost.

> Something that I am also thinking is to completely drop `sc_putc()` and require on every lower-half driver to implement an `sc_write()` only, with the requirement that the driver will perform the write atomically. If the driver cannot actually perform a write, it can iterate internally in a protected section of the code.

Yes, I also think sc_putc make the interface more complicated than necessary. Maybe, it's better to remove sc_putc first before we redesign the syslog channel.

Reply:
@xiaoxiang781216 If I understand this correctly, `struct syslog_dev_s` should be moved to `include/nuttx/syslog/syslog.h`.
And then `struct syslog_channel_s` will be something like:

```c
struct syslog_channel_s
{
  /* Channel device */

  FAR struct syslog_dev_s *sc_dev;

  /* Channel operations */

  FAR struct syslog_channel_ops_s *ops;

  /* Implementation specific logic may follow */
};
```

Correct?
Or should `sc_dev` remain a `void *`, and keep `struct syslog_dev_s *sc_dev` private?